README.pod
=pod

=encoding utf8

=head1 NAME

Mojo::IOLoop::Subprocess::Sereal - Subprocesses with Sereal

=head1 SYNOPSIS

  use Mojo::IOLoop::Subprocess::Sereal;

  # Operation that would block the event loop for 5 seconds
  my $subprocess = Mojo::IOLoop->$_subprocess(
    sub {
      my $subprocess = shift;
      sleep 5;
      return '♥', 'Mojolicious';
    },
    sub {
      my ($subprocess, $err, @results) = @_;
      say "I $results[0] $results[1]!";
    }
  );

  # Start event loop if necessary
  Mojo::IOLoop->start unless Mojo::IOLoop->is_running;

=head1 DESCRIPTION

L<Mojo::IOLoop::Subprocess::Sereal> provides a L</"$_subprocess"> method which
works as a drop-in replacement for L<Mojo::IOLoop/"subprocess"> while using
L<Sereal> for data serialization. L<Sereal> is faster than L<Storable> and
supports serialization of more reference types such as C<Regexp>. The
L<Sereal::Encoder/"FREEZE/THAW CALLBACK MECHANISM"> is supported to control
serialization of blessed objects.

See L<Mojo::IOLoop::Subprocess::Role::Sereal> for a role to apply L<Sereal>
data serialization to any L<Mojo::IOLoop::Subprocess>.

=head1 EXPORTS

L<Mojo::IOLoop::Subprocess::Sereal> exports the following variable by default.

=head2 $_subprocess

  my $subprocess = Mojo::IOLoop->$_subprocess(sub {...}, sub {...});
  my $subprocess = Mojo::IOLoop->$_subprocess;
  my $subprocess = $loop->$_subprocess(sub {...}, sub {...});

Build L<Mojo::IOLoop::Subprocess> object to perform computationally expensive
operations in subprocesses, without blocking the event loop. Composes and calls
L<Mojo::IOLoop::Subprocess::Role::Sereal/"with_sereal"> to use L<Sereal> for
data serialization. If arguments are provided, they will be used to call
L<Mojo::IOLoop::Subprocess/"run">.

=head1 BUGS

Report any issues on the public bugtracker.

=head1 AUTHOR

Dan Book <dbook@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2016 by Dan Book.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=head1 SEE ALSO

L<Mojo::IOLoop>, L<Mojo::IOLoop::Subprocess::Role::Sereal>, L<Sereal>

=cut


Changes
{{$NEXT}}

1.002     2021-06-15 23:59:14 EDT
  - Update IRC metadata to libera.chat

1.001     2021-02-16 17:19:49 EST
  - Use Mojo::Promise in tests instead of the deprecated and decored Mojo::IOLoop::Delay.

1.000     2017-12-17 17:05:24 EST
  - Add Mojo::IOLoop::Subprocess::Role::Sereal and use it to provide the $_subprocess method.
  - Update $_subprocess to not call ->run if given no callbacks (as in Mojolicious 7.59).

0.005     2017-03-21 01:17:50 EDT
  - Include Sereal::Encoder and Sereal::Decoder versions in tests for debugging purposes

0.004     2016-12-28 22:25:15 EST
  - Remove experimental designation from module.

0.003     2016-09-27 01:55:58 EDT
  - Remove subclass interface in favor of exported $_subprocess method
    (This module does say EXPERIMENTAL!)
  - Export $_subprocess method by default

0.002     2016-09-27 00:34:55 EDT
  - Export $_subprocess method as drop-in for Mojo::IOLoop->subprocess()

0.001     2016-09-19 21:44:22 EDT
  - First release


lib/Mojo/IOLoop/Subprocess/Sereal.pm
package Mojo::IOLoop::Subprocess::Sereal;

use strict;
use warnings;
use Exporter 'import';

our $VERSION = '1.003';

our @EXPORT = '$_subprocess';

our $_subprocess = sub {
  my $subprocess = shift->subprocess
    ->with_roles('Mojo::IOLoop::Subprocess::Role::Sereal')->with_sereal;
  return @_ ? $subprocess->run(@_) : $subprocess;
};

1;

=encoding utf8

=head1 NAME

Mojo::IOLoop::Subprocess::Sereal - Subprocesses with Sereal

=head1 SYNOPSIS

  use Mojo::IOLoop::Subprocess::Sereal;

  # Operation that would block the event loop for 5 seconds
  my $subprocess = Mojo::IOLoop->$_subprocess(
    sub {
      my $subprocess = shift;
      sleep 5;
      return '♥', 'Mojolicious';
    },
    sub {
      my ($subprocess, $err, @results) = @_;
      say "I $results[0] $results[1]!";
    }
  );

  # Start event loop if necessary
  Mojo::IOLoop->start unless Mojo::IOLoop->is_running;

=head1 DESCRIPTION

L<Mojo::IOLoop::Subprocess::Sereal> provides a L</"$_subprocess"> method which
works as a drop-in replacement for L<Mojo::IOLoop/"subprocess"> while using
L<Sereal> for data serialization. L<Sereal> is faster than L<Storable> and
supports serialization of more reference types such as C<Regexp>. The
L<Sereal::Encoder/"FREEZE/THAW CALLBACK MECHANISM"> is supported to control
serialization of blessed objects.

See L<Mojo::IOLoop::Subprocess::Role::Sereal> for a role to apply L<Sereal>
data serialization to any L<Mojo::IOLoop::Subprocess>.

=head1 EXPORTS

L<Mojo::IOLoop::Subprocess::Sereal> exports the following variable by default.

=head2 $_subprocess

  my $subprocess = Mojo::IOLoop->$_subprocess(sub {...}, sub {...});
  my $subprocess = Mojo::IOLoop->$_subprocess;
  my $subprocess = $loop->$_subprocess(sub {...}, sub {...});

Build L<Mojo::IOLoop::Subprocess> object to perform computationally expensive
operations in subprocesses, without blocking the event loop. Composes and calls
L<Mojo::IOLoop::Subprocess::Role::Sereal/"with_sereal"> to use L<Sereal> for
data serialization. If arguments are provided, they will be used to call
L<Mojo::IOLoop::Subprocess/"run">.

=head1 BUGS

Report any issues on the public bugtracker.

=head1 AUTHOR

Dan Book <dbook@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2016 by Dan Book.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=head1 SEE ALSO

L<Mojo::IOLoop>, L<Mojo::IOLoop::Subprocess::Role::Sereal>, L<Sereal>


Makefile.PL
# This Makefile.PL for Mojo-IOLoop-Subprocess-Sereal was generated by
# Dist::Zilla::Plugin::MakeMaker::Awesome 0.47.
# Don't edit it but the dist.ini and plugins used to construct it.

use strict;
use warnings;

use 5.010001;
use ExtUtils::MakeMaker;

use Config;
if ($Config{d_pseudofork}) {
  print "Subprocesses not supported with fork emulation\n";
  exit 0;
}

my %WriteMakefileArgs = (
  "ABSTRACT" => "Subprocesses with Sereal",
  "AUTHOR" => "Dan Book <dbook\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Mojo-IOLoop-Subprocess-Sereal",
  "LICENSE" => "artistic_2",
  "MIN_PERL_VERSION" => "5.010001",
  "NAME" => "Mojo::IOLoop::Subprocess::Sereal",
  "PREREQ_PM" => {
    "Exporter" => 0,
    "Mojolicious" => "7.59",
    "Role::Tiny" => "2.000001",
    "Sereal" => "3.001"
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => "0.88"
  },
  "VERSION" => "1.001",
  "test" => {
    "TESTS" => "t/*.t"
  }
);

my %FallbackPrereqs = (
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Mojolicious" => "7.59",
  "Role::Tiny" => "2.000001",
  "Sereal" => "3.001",
  "Test::More" => "0.88"
);

unless ( eval { ExtUtils::MakeMaker->VERSION('6.63_03') } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
